### PR TITLE
Best practices: design for mobile experience

### PR DESCRIPTION
**Pull request by catong**
This PR adds a new section to the "Building and Running an edX Course" guide for mobile best practices. This initial draft is fairly basic, just to get some of this info into the guide. Ultimately the goal is to rewrite all sections of the guide so that "designing for mobile" isn't a separate concept but is well incorporated throughout the documentation. 
https://openedx.atlassian.net/browse/DOC-2117
### Reviewers

Check the appropriate box after each reviewer finishes and gives :+1:.
- [ ] Subject matter expert: @nasthagiri 
- [x] Subject matter expert: @aleffert 
- [x] UX team review: @frrrances 
- [x] Doc team review: @lamagnifica  
- [x] Product review: @explorerleslie 
- FYI: @marcotuts @lou-wang 
### Draft HTML Documentation
- [x] http://draft-br-reach-learners.readthedocs.org/en/latest/reaching_learners/design_for_mobile.html
### Post-review
- [ ] Squash commits

### CODE REVIEW

**Comment by nasthagiri** [comment]
Please include someone from the UX team (@frrrances) and/or @marcotuts as a subject matter expert reviewer.  They can give advice on creating responsive HTML, especially about image rendering, setting HTML width sizes, and creating touch-sized controls.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+* Be concise with display names and labels. Long display names and labels
+  might wrap on smaller screens, or might not be easily viewable. For example,
+  if several components have names that all start with the first five words
+  and differ only after that, learners on mobile devices will have a difficult
+  time distinguishing between components.
+
+* Do not assume that learners are viewing your course materials on a
+  particular screen size. Optimize your materials for viewing on a range of
+  screen sizes. If you upload images, provide a high resolution version that
+  can be zoomed in for viewing.
```
Comment (aleffert) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: Not sure this follows? Typically you want smaller images that fit better into tighter layouts.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+* When you make choices about the problem types to use for graded versus
+  ungraded assignments in your course, or which problem types to present in a
+  single unit, keep the mobile experience in mind. It is a best practice to
+  use only mobile-ready problem types for graded assignments.
```
Comment (aleffert) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: Are we ready to say this? I feel like our support of assessments is still kind of experimental as is?

**Comment by aleffert** [comment]
That's all I have. Looks good :+1:

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 1 above)
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: I would reword this into two separate points (wordsmith away, I'm just spitballing): 
- Remember that your course content may be viewed on a screen as large as a high-resolution Mac Thunderbolt retina-enabled display to an iPhone 4s with a very small screen. Thus, it can be difficult to size an image that displays well at all resolutions. We recommend keeping images most images under 0.5MB for quick downloading for low-bandwidth learners.
- If you have a large image that requires zooming to view the full detail, link to a full-resolution copy that can be opened separately from the courseware.

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 2 above)
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: Yeah, not ready to say this. I would frame it as something like, "wherever possible, use mobile friendly assessment types. If you're using unsupported assessment types, give your students a head's up in the courseware that they won't be able to complete that assignment on mobile."

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+* Embedded images can be problematic. When images are rendered on different
+  displays, they might display in an unintended way, for example if the aspect
+  ratio is changed.
```
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: I don't think embedded images are problematic. I think it's one of the most helpful things that can be included in courses, to illustrate concepts. I think they just have to be styled in a mobile friendly way. @frrrances can you comment here on advice for styling images so they render on mobile well? I _think_ in my testing that setting width=100% helped a lot.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+.. _Testing Your Course For Mobile Devices:
+
+**************************************
+Testing Your Course For Mobile Devices
+**************************************
+
+If you have included some of the more complex problem types, or have
+customized any of the course content, edX recommends that you test your course
```
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: maybe "highly customized the display of course content" -- everyone customizes course content in some way, it's only the people who go hog-wild who need to test thoroughly.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+for multiple devices and displays.
+
+To test the mobile experience or your course, log in to your course on a
+mobile device, using a browser such as Chrome. The courseware components will
+look in the browser the way that they will look in the mobile app, although
+the browser header and navigation pane will not be exactly the same.
```
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: I don't know if this is a great way to test. The heading and nav get totally in the way of rendering, and the content isn't responsive. Why don't we just tell them to preview / test in the app itself? If staff logs in, they'll be able to see both released and unreleased content that has been published, same as on the desktop.

**Comment by explorerleslie** [comment]
@catong finished first pass review and left a bunch of comments. Also, can we add recommendations about not using Flash, which doesn't render on mobile, and limiting the use of iFrames to only content that must be iFrame'd into the course, because they normally aren't as responsive as content built in Studio. I also agree that having @frrrances take a look and make recommendations for HTML / CSS dos and don'ts would be a good idea -- Frances, can you have a look? Please retag me when ready for review!

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 5 above)
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: +1 on Leslie's comment -- the browser experience on mobile devices should be quite similar to the browser experience on a desktop, just impossibly small and hard to navigate...

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 6]
```diff
@@ -0,0 +1,71 @@
+.. _Designing For a Mobile Experience:
+
+###############################################
+Designing Your Course For a Mobile Experience
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst:4]: a --> the ?

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+###############################################
+
+The percentage of learners who access MOOCs using a mobile device is
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: a mobile device --> their mobile devices ? should be plural (from learners), but is increasing should stay singular (from percentage) I think... ?

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+increasing every day. Most courses on edX.org can be viewed on mobile devices,
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: edx.org all lower case

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+although we still recommend that learners complete graded assignments on a
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: we --> edX

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+desktop computer, depending on the type of assessments their courses include.
+For information on which exercises and tools are mobile-ready, see the table
+in the :ref:`Introduction to Exercises and Tools<Create Exercises>` section.
+
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst:14]: A semantic note: Lou had strong feelings about how we referred to the experience of using the edX mobile apps (only available on Android smartphones and iPhones)  as differentiated from the experience of using a browser to access courseware (could be on any type of computer, not just desktop computers, and also could be from tablets, phablets, smartphones and iPhones). I'm not sure if @explorerleslie has as specific and definitive a terminology in mind, but that was at one time such a strong preference that it's in the style guide.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  include an HTML component in a unit for the purpose of providing a
+  demarcation point or guiding learners to the next unit, but having to swipe
+  through too many "markers" with no real course content is not a good
+  experience for mobile users.
+
+* Display names are critical for navigation on mobile devices. As you create
+  course content, make sure you replace the default display names for every
+  component with useful courseware component names.
+
+* Be concise with display names and labels. Long display names and labels
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: maybe, Make display names and labels concise. ? or, Keep display names... ?

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  if several components have names that all start with the first five words
+  and differ only after that, learners on mobile devices will have a difficult
+  time distinguishing between components.
+
+* Do not assume that learners are viewing your course materials on a
+  particular screen size. Optimize your materials for viewing on a range of
+  screen sizes. If you upload images, provide a high resolution version that
+  can be zoomed in for viewing.
+
+* When you make choices about the problem types to use for graded versus
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: versus could just be "and" -- they're not really in opposition to each other

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  single unit, keep the mobile experience in mind. It is a best practice to
+  use only mobile-ready problem types for graded assignments.
+
+* Embedded images can be problematic. When images are rendered on different
+  displays, they might display in an unintended way, for example if the aspect
+  ratio is changed.
+
+* If you make JavaScript or CSS changes that affect more than one component
+  (in the XML), be aware that the component might not render in the same way
+  on a mobile device as it will in the desktop. EdX recommends that you create
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: maybe, If you make changes to the JavaScript or CSS by editing your course's XML, be aware that the component ...
(or just reword to get agreement for "more than one component... the component")

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+**************************************
+Testing Your Course For Mobile Devices
+**************************************
+
+If you have included some of the more complex problem types, or have
+customized any of the course content, edX recommends that you test your course
+for multiple devices and displays.
+
+To test the mobile experience or your course, log in to your course on a
```
Comment (lamagnifica) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: log in --> sign in

**Comment by lamagnifica** [comment]
thanks for getting this going! :+1: with some suggestions and questions

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 10 above)
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst:14]: Yes, I still agree that wherever possible, "edX mobile apps" should be changed to "edX Android smartphone and iPhone apps".

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 5 above)
Comment (catong) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: These must be outdated recommendations that were captured in the JIRA story; I wasn't familiar enough with Mobile to be sure this didn't make sense!

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 6 above)
Comment (catong) [inline en_us/shared/reaching_learners/design_for_mobile.rst:4]: I don't feel that this needs to read as if there is only one possible mobile experience that we are referring specifically to; I'm inclined to stick with "a" unless this reads strangely to everyone else.

**Comment by catong** [comment]
@frrrances Do you have time to take a look and provide any HTML/CSS best practices/warnings?

**Comment by catong** [comment]
@frrrances Do you have time to take a look and provide some HTML/CSS best practices/warnings?
@explorerleslie FYI I've updated per review comments to date, but you might want to wait to see if there is new content coming from Frances. Feel free to ignore for now; I'll tag you again.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  resolutions. In general, edX recommends keeping most images under 0.5MB in
+  size so that learners who have low Internet bandwidth will not have trouble
+  downloading the images. If you have a large image that requires zooming to
+  view the full detail, in addition to providing an image that can be easily
+  downloaded, link to a full resolution copy that can be opened separately
+  from the courseware.
+
+* When you make choices about the problem types to use for graded and ungraded
+  assignments in your course, or which problem types to present in a single
+  unit, keep the mobile experience in mind. Whenever possible, use mobile-
```
Comment (frrrances) [inline en_us/shared/reaching_learners/design_for_mobile.rst:53]: Is there a place that lists the mobile-friendly assessments? It would be nice to link to there from here.

**Comment by frrrances** [comment]
Looks good, @catong! I think what you have here will be the most important things for authors to keep in mind. The LMS and mobile apps should handle most of the styles on smaller screens for them.

I'm not sure how into the details you want to get with implementation, but if they might be editing the code of say an HTML block, you might warn people not to explicitly set sizes of things like text, tables, images, etc as they won't scale as well for mobile. This seems to happen most often in the Course Announcements, though I've seen it in content too. Ideally, they should use the CodeMirror tools to make their content semantic (headers, paragraphs, etc), but really that would help so many things (accessibility, content hierarchy, style consistency across the app and various contexts).

:+1:

File: en_us/shared/reaching_learners/design_for_mobile.rst (hunk 15 above)
Comment (catong) [inline en_us/shared/reaching_learners/design_for_mobile.rst:53]: @frrrances There is, but I did not include it at this stage because mobile assessments has not been released. Thanks for the reminder; I'll add a link that can be commented out or updated, depending on whether these best practices still get published before assessments are released.

**Comment by catong** [comment]
Thanks, @frrrances! I added a bullet item about relative instead of explicit sizing in HTML per your review.
@explorerleslie Ready for your 2nd review.

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+  useful to include an HTML component in a unit for the purpose of providing a
+  demarcation point or guiding learners to the next unit, but having to swipe
+  through too many "markers" with no real course content is not a good
+  experience for mobile users.
+
+* If you make JavaScript or CSS changes to course components using XML, be
+  aware that components might not render in the same way on a mobile device as
+  they do in the desktop. EdX recommends that you create and edit components
+  in Studio rather than editing them in XML, to ensure that components render
+  reliably on both mobile and desktop.
```
Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: I'm not sure that this one matters. If you know what you're doing, it shouldn't matter whether you edit CSS or Javascript in XML or in Studio. I also don't want to sound like we're discouraging use of OLX. Could we soften this to something like, when creating components in XML, ensure your JS and CSS are compliant, and verify in LMS,  because you don't have the benefit of live preview in Studio, or something?

File: en_us/shared/reaching_learners/design_for_mobile.rst [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+  progress from component to component by swiping through them. It might seem
+  useful to include an HTML component in a unit for the purpose of providing a
+  demarcation point or guiding learners to the next unit, but having to swipe
+  through too many "markers" with no real course content is not a good
+  experience for mobile users.
+
+* If you create course components using XML, make sure your JavaScript and CSS
+  are compliant. You should verify that all components render correctly in the
+  edX Android and iPhone apps as well as directly in the LMS, because previewing
+  from Studio is not available for components created outside of Studio
```
Comment (catong) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: @explorerleslie -- is this better?

Comment (explorerleslie) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: Gah, sorry. So preview is available for XML authored components in most cases. Maybe we just leave this as: 

Make sure your JavaScript and CSS are compliant. You should verify that all components render correctly in the edX Android and iPhone apps as well as directly in the LMS on desktop.

Comment (catong) [inline en_us/shared/reaching_learners/design_for_mobile.rst]: @explorerleslie Made your suggested updates; thanks!. Please give thumbs if you think this is ready to go.

**Comment by explorerleslie** [comment]
:+1:

**Comment by catong** [comment]
Squashed commits; merging https://github.com/edx/edx-documentation/pull/593
